subprojects/docs/src/docs/userguide/extending-gradle/lazy_configuration.adoc

```diff
@@ -277,14 +277,24 @@ $ gradle show
 include::{samplesPath}/providers/propertyConvention/propertyConvention.out[]
 ----
 
-[[sec:unmodifiable_property]]
+[[unmodifiable_property]]
 == Making a property unmodifiable
 
-Most properties of a task or project are intended to be configured by plugins or build scripts and then the resulting value used to do something useful. For example, a property that specifies the output directory for a compilation task may start off with a value specified by a plugin, then a build script might configure the value to some custom location, then this value is used by the task when it runs. However, once the task starts to run, we want to prevent any further change to the property. This way we avoid errors that result from different consumers, such as the task action or Gradle's up-to-date checks or build caching or other tasks, using different values for the property.
+Most properties of a task or project are intended to be configured by plugins or build scripts and then the resulting value used to do something useful. For example, a property that specifies the output directory for a compilation task may start off with a value specified by a plugin, then a build script might change the value to some custom location, then this value is used by the task when it runs.
+However, once the task starts to run, we want to prevent any further change to the property. This way we avoid errors that result from different consumers, such as the task action or Gradle's up-to-date checks or build caching or other tasks, using different values for the property.
 
-Lazy properties provide a link:{javadocPath}/org/gradle/api/provider/Property.html#finalizeValue--[finalizeValue()] method to make this explicit. Calling this method makes a property instance unmodifiable from that point on and any further attempts to change the value of the property will fail. Gradle automatically makes the properties of a task final when the task starts execution.
+Lazy properties provide several methods that you can use to disallow changes to their value once the value has been configured.
+The link:{javadocPath}/org/gradle/api/provider/Property.html#finalizeValue--[finalizeValue()] method calculates the _final_ value for the property and prevents further changes to the property.
+When the value of the property comes from a `Provider`, the provider is queried for its current value and the result becomes the final value for the property.
+This final value replaces the provider and the property no longer tracks the value of the provider.
+Calling this method also makes a property instance unmodifiable and any further attempts to change the value of the property will fail.
```

Comment:
:o: This is what I meant, this behavior should be documented in the Javadocs too.